Store Orders: Enable "Edit Order" button

ryelle opened the pull request:
This PR adds a button to toggle on the "edit state" of a given order. Right now the only thing that can be changed in the edit state is the order status. Since this page will also be the create screen, I've re-used the feature flag `woocommerce/extension-orders-create`, which is enabled only in `development` & `wp-calypso`.

<img width="855" alt="flag-enabled" src="https://user-images.githubusercontent.com/541093/31148313-5eb31998-a85a-11e7-9aa8-e8be0b9e1a47.png">

One you're in the edit state, you can change the order status. The save button is disabled until changes are made.

<img width="837" alt="screen shot 2017-10-03 at 5 07 28 pm" src="https://user-images.githubusercontent.com/541093/31149282-5dd238bc-a85d-11e7-95ff-0ca6a0f045af.png">
<img width="845" alt="screen shot 2017-10-03 at 5 05 37 pm" src="https://user-images.githubusercontent.com/541093/31149253-4c48d5e2-a85d-11e7-84e5-426d585eff5b.png">

The "Order created…", payment card, and fulfillment card are not shown in the "edit state" to reduce the complexity of the screen.

For staging & production, where the feature flag is not enabled, users will not be able to edit the order, including changing the order status using the dropdown. Users can mark as paid, refund, or fulfill orders using the actions in those sections, so there should be no issue while editing is being developed. Screenshot:

<img width="855" alt="flag-disabled" src="https://user-images.githubusercontent.com/541093/31148315-5ec06440-a85a-11e7-9c46-66900e5e5b72.png">

**To test**

- View an order by going to `/store/orders/:site` and clicking into one
- You should see an "Edit Order" button in the top right corner
- Click it to enable edit state, a disabled "Save Order" button appears in its place
- Change the order status, the save button should switch to enabled
- Click save, it should switch you out of edit mode, and show a green "Order saved" notice.
- Repeat, cancelling out of an edit without making changes, or with changes pending (that will be cleared without saving)

Test with the flag disabled by running `DISABLE_FEATURES=woocommerce/extension-orders-create npm start` to make sure nothing is broken without the flag.

timmyc commented on client/extensions/woocommerce/app/order/index.js:
```diff
@@ -42,43 +45,90 @@ class Order extends Component {
-		this.props.updateOrder( this.props.siteId, this.state.order );
-	}
+		const { siteId, order } = this.props;
+		this.props.updateOrder( siteId, order );
```
Do we need to consider the possibility of the `updateOrder` API request failing before we `clearOrderEdits` in the subsequent line?

timmyc commented on client/extensions/woocommerce/app/order/index.js:
```diff
@@ -42,43 +45,90 @@ class Order extends Component {
+			</span>,
 		];
+
+		let button = [
```
Heh, I was going to say maybe this should be _buttons_ but see that it can be singular button below. 6 one way, half dozen the other.

timmyc commented on client/extensions/woocommerce/app/order/index.js:
```diff
@@ -42,43 +45,90 @@ class Order extends Component {
+		}
+	};
 
+	// Clear this order's edits, takes it out of edit state
```
Since the orderEdit state is only tied to a `siteId` and not an `orderId` - is there any chance we could get a polluted `orderEdit` state if someone edits an order here, then closes out the tab, or never executes a save or cancel action?

ryelle commented:
Thanks for testing the conflict, @timmyc – I've fixed that by added `clearOrderEdits` to `componentWillUnmount` and `componentWillReceiveProps`, for when the displayed order changes. That should clear the conflicts between orders, and acts like I would expect: moving away from the edits screen erases edits -- the only thing here is maybe an are-you-sure intercept if there are pending changes (but I can do that in a following PR).

I agree that we should wait to see if the update succeeds before clearing edits, so I've moved that task to the edits reducer-- now if the edits reducer sees a successful update, it clears the pending edit state.
